qiskit/dagcircuit/dagnode.py

```diff
@@ -12,16 +12,146 @@
 
 # pylint: disable=redefined-builtin
 
-"""Object to represent the information at a node in the DAGCircuit."""
+"""Objects to represent the information at a node in the DAGCircuit."""
 
 import warnings
 
 from qiskit.exceptions import QiskitError
 
 
+class DAGNodeP:
+    """Parent class for OpNode, InNode, and OutNode used by DAGCircuit."""
```

Comment:
I wonder if instead of creating a new class, we could instrument something with `__new__` such that users calling `DAGNode(type='op', ...)` get both a `DeprecationWarning` and an instance of `OpNode` (which should behave identically to the current `DAGNode(type='op', ...)`. Something like:
```
class DAGNode:
  def __new__(cls, *args, **kwargs):
    if ((args and args[0] == 'op') or (kwargs and kwargs['type'] == 'op'):
      warnings.warn(...)
      return super().__new__(OpNode, *args, **kwargs)
    elif (...):
      ...
```

Reply:
I'm not sure why a user would be creating a DAGNode. All of the user facing functions in `dagcircuit` like `apply_operation_back` use op instead of node.

I also thought about creating an OpNode when DAGNode is called, but you'd be changing the interface to the node, so something like `if type == 'op'` wouldn't work with OpNode. That's why I thought it best to have a clean parent function for the new stuff and then put the deprecations in the old DAGNode. 

If someone is creating they're own DAG-likeCircuits using qiskit DAGNodes (again not sure why), the DAGNode would continue to work the same. Once the deprecation period is finished, the whole class could then be deleted.